telegram/ext/builders.py

```diff
@@ -0,0 +1,1196 @@
+    ('update_queue', 'update_queue'),
+    ('workers', 'workers'),
+    ('exception_event', 'exception_event'),
+    ('job_queue', 'JobQueue instance'),
+    ('persistence', 'persistence instance'),
+    ('context_types', 'ContextTypes instance'),
+    ('dispatcher_class', 'Dispatcher Class'),
+] + _BOT_CHECKS
+_DISPATCHER_CHECKS.remove(('dispatcher', 'Dispatcher instance'))
+
+_TWO_ARGS_REQ = "The parameter `{}` may only be set, if no {} was set."
+
+
+# Base class for all builders. We do this mainly to reduce code duplication, because e.g.
+# the UpdaterBuilder has all method that the DispatcherBuilder has
+class _BaseBuilder(Generic[ODT, BT, CCT, UD, CD, BD, JQ, PT]):
+    # pylint reports false positives here:
+    # pylint: disable=W0238
+
+    __slots__ = (
+        '_token',
+        '_base_url',
+        '_base_file_url',
+        '_request_kwargs',
+        '_request',
+        '_private_key',
+        '_private_key_password',
+        '_defaults',
+        '_arbitrary_callback_data',
+        '_bot',
+        '_update_queue',
+        '_workers',
+        '_exception_event',
+        '_job_queue',
+        '_persistence',
+        '_context_types',
+        '_dispatcher',
+        '_user_signal_handler',
+        '_dispatcher_class',
+        '_dispatcher_kwargs',
+        '_updater_class',
+        '_updater_kwargs',
+    )
+
+    def __init__(self: 'InitBaseBuilder'):
+        self._token: DVInput[str] = DefaultValue('')
+        self._base_url: DVInput[str] = DefaultValue('https://api.telegram.org/bot')
+        self._base_file_url: DVInput[str] = DefaultValue('https://api.telegram.org/file/bot')
+        self._request_kwargs: DVInput[Dict[str, Any]] = DefaultValue({})
+        self._request: ODVInput['Request'] = DEFAULT_NONE
+        self._private_key: ODVInput[bytes] = DEFAULT_NONE
+        self._private_key_password: ODVInput[bytes] = DEFAULT_NONE
+        self._defaults: ODVInput['Defaults'] = DEFAULT_NONE
+        self._arbitrary_callback_data: DVInput[Union[bool, int]] = DEFAULT_FALSE
+        self._bot: Bot = DEFAULT_NONE  # type: ignore[assignment]
+        self._update_queue: DVInput[Queue] = DefaultValue(Queue())
+        self._workers: DVInput[int] = DefaultValue(4)
+        self._exception_event: DVInput[Event] = DefaultValue(Event())
+        self._job_queue: ODVInput['JobQueue'] = DefaultValue(JobQueue())
+        self._persistence: ODVInput['BasePersistence'] = DEFAULT_NONE
+        self._context_types: DVInput[ContextTypes] = DefaultValue(ContextTypes())
+        self._dispatcher: ODVInput['Dispatcher'] = DEFAULT_NONE
+        self._user_signal_handler: Optional[Callable[[int, object], Any]] = None
+        self._dispatcher_class: DVInput[Type[Dispatcher]] = DefaultValue(Dispatcher)
+        self._dispatcher_kwargs: Dict[str, object] = {}
+        self._updater_class: Type[Updater] = Updater
+        self._updater_kwargs: Dict[str, object] = {}
+
+    def _get_connection_pool_size(self) -> int:
```

Comment:
This might be my personal taste, but I prefer such a function to be a staticmethod and have it require as an input the workers count.
I see the following benefits:
1. It allows more flexibility for the consumer, passing a different worker count, depending on variables of its own control.
2. This approach makes it easier adding a unitest which tests the logic of this function, and this function alone (and maybe we should add  such a unitest)

i.e.
```python
def _get_connection_pool_size(workers: int) -> int:
    return workers + 4
```